In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt  
%matplotlib inline        
import pandas as pd
import pycaret
from pycaret.regression import setup, compare_models


In [2]:
!pip install openpyxl
!pip install xgboost

### 1. Pycaret sin imputar valores ni hacer OHE

In [3]:
df_pycaret = pd.read_excel(r"C:\Users\evahr\Downloads\TFM-idealista\df_lag_pycaret.xlsx")

df_pycaret

AÑO  MES_NUM    DISTRITO   LATITUD  LONGITUD    SHAPE_AREA  PRECIO M2  \
0     2007       11  Arganzuela  40.40021  -3.69618  6.462176e+06       4000   
1     2007       12  Arganzuela  40.40021  -3.69618  6.462176e+06       4000   
2     2008        1  Arganzuela  40.40021  -3.69618  6.462176e+06       4000   
3     2008        2  Arganzuela  40.40021  -3.69618  6.462176e+06       3824   
4     2008        3  Arganzuela  40.40021  -3.69618  6.462176e+06       4000   
...    ...      ...         ...       ...       ...           ...        ...   
4413  2024       12  Villaverde  40.35000  -3.70000  2.019432e+07       2252   
4414  2025        1  Villaverde  40.35000  -3.70000  2.019432e+07       2207   
4415  2025        2  Villaverde  40.35000  -3.70000  2.019432e+07       2254   
4416  2025        3  Villaverde  40.35000  -3.70000  2.019432e+07       2294   
4417  2025        4  Villaverde  40.35000  -3.70000  2.019432e+07       2362   

      EURIBOR_lag12  VARIACIÓN ANUAL_lag12  PRECIO M2_lag12  \
0               NaN                    NaN              NaN   
1               NaN                    NaN              NaN   
2               NaN                    NaN              NaN   
3               NaN                    NaN              NaN   
4               NaN                    NaN              NaN   
...             ...                    ...              ...   
4413        0.03679                  0.063           1963.0   
4414        0.03609                  0.068           1982.0   
4415        0.03671                  0.058           1977.0   
4416        0.03716                  0.067           1988.0   
4417        0.03702                  0.085           2046.0   

      TRANSACCIONES_CM_lag12  INDICE_PRECIO_lag12  TOTAL_HIPOTECAS_lag12  \
0                        NaN                  NaN                    NaN   
1                        NaN                  NaN                    NaN   
2                        NaN                  NaN                    NaN   
3                        NaN                  NaN                    NaN   
4                        NaN                  NaN                    NaN   
...                      ...                  ...                    ...   
4413                  4048.0               164.82                 6548.0   
4414                  6209.0               168.79                 8547.0   
4415                  6336.0               168.79                 8492.0   
4416                  5924.0               168.79                 7033.0   
4417                  5805.0               174.54                 7177.0   

      POBLACION_ACTIVA_lag12  POBLACION_lag12  ESPERANZA_VIDA_lag12  \
0                        NaN              NaN                   NaN   
1                        NaN              NaN                   NaN   
2                        NaN              NaN                   NaN   
3                        NaN              NaN                   NaN   
4                        NaN              NaN                   NaN   
...                      ...              ...                   ...   
4413                  3720.2         318076.0                  85.9   
4414                  3750.8         335904.0                  88.8   
4415                  3750.8         335904.0                  88.8   
4416                  3750.8         335904.0                  88.8   
4417                  3749.9         335904.0                  88.8   

      VIVIENDAS_COMPRAVENTA_lag12  TRANSACCIONES_SUELO_lag12  \
0                             NaN                        NaN   
1                             NaN                        NaN   
2                             NaN                        NaN   
3                             NaN                        NaN   
4                             NaN                        NaN   
...                           ...                        ...   
4413                        815.0                      345.0   
4414       

In [4]:
setup(data=df_pycaret, target='PRECIO M2', preprocess=False, session_id=123)

In [5]:
from pycaret.regression import setup, compare_models, pull

# Setup inicial
exp = setup(
    data=df_pycaret,
    target="PRECIO M2",
    session_id=123,
    normalize=True,
    remove_multicollinearity=True,
    multicollinearity_threshold=0.8,
    imputation_type="simple",
    numeric_imputation="median",
    categorical_imputation="mode",
    use_gpu=True,
    fold_strategy="timeseries",
    data_split_shuffle=False,   
    fold_shuffle=False
)

# Comparación de modelos
best_models = compare_models(n_select=10, sort="R2")


[LightGBM] [Warning] There are no meaningful features which satisfy the provided configuration. Decreasing Dataset parameters min_data_in_bin or min_data_in_leaf and re-constructing Dataset might resolve this warning.
[LightGBM] [Info] This is the GPU trainer!!
[LightGBM] [Info] Total Bins 0
[LightGBM] [Info] Number of data points in the train set: 2, number of used features: 0
[LightGBM] [Info] Using GPU Device: NVIDIA GeForce RTX 3050 OEM, Vendor: NVIDIA Corporation
[LightGBM] [Info] Compiling OpenCL Kernel with 16 bins...
[LightGBM] [Info] GPU programs have been built
[LightGBM] [Warning] GPU acceleration is disabled because no non-trivial dense features can be found
[LightGBM] [Info] Start training from score 0.500000
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[LightGBM] [Warning] Stopped training because there are no m

[LightGBM] [Warning] There are no meaningful features which satisfy the provided configuration. Decreasing Dataset parameters min_data_in_bin or min_data_in_leaf and re-constructing Dataset might resolve this warning.
[LightGBM] [Info] This is the GPU trainer!!
[LightGBM] [Info] Total Bins 0
[LightGBM] [Info] Number of data points in the train set: 2, number of used features: 0
[LightGBM] [Info] Using GPU Device: NVIDIA GeForce RTX 3050 OEM, Vendor: NVIDIA Corporation
[LightGBM] [Info] Compiling OpenCL Kernel with 16 bins...
[LightGBM] [Info] GPU programs have been built
[LightGBM] [Warning] GPU acceleration is disabled because no non-trivial dense features can be found
[LightGBM] [Info] Start training from score 0.500000
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[LightGBM] [Warning] Stopped training because there are no m

In [6]:
# Extraer resultados de la última operación (compare_models)
results = pull()

# Mostrar tabla (DataFrame con R2, RMSE, MAE, etc.)
results


Model           MAE           MSE  \
omp           Orthogonal Matching Pursuit  2.214173e+02  1.061940e+05   
gbr           Gradient Boosting Regressor  2.592703e+02  1.460628e+05   
xgboost         Extreme Gradient Boosting  2.750347e+02  1.565597e+05   
lasso                    Lasso Regression  3.102573e+02  1.675305e+05   
rf                Random Forest Regressor  2.854094e+02  1.725915e+05   
llar         Lasso Least Angle Regression  3.148613e+02  1.725818e+05   
et                  Extra Trees Regressor  3.298491e+02  1.957281e+05   
ridge                    Ridge Regression  3.273401e+02  1.820649e+05   
br                         Bayesian Ridge  3.276265e+02  1.824755e+05   
lightgbm  Light Gradient Boosting Machine  2.849510e+02  1.959110e+05   
ada                    AdaBoost Regressor  3.323053e+02  2.027365e+05   
dt                Decision Tree Regressor  3.436374e+02  2.250874e+05   
par          Passive Aggressive Regressor  3.324782e+02  1.926046e+05   
huber                     Huber Regressor  3.387426e+02  2.010974e+05   
en                            Elastic Net  5.347201e+02  4.419643e+05   
knn                 K Neighbors Regressor  6.144785e+02  6.026311e+05   
dummy                     Dummy Regressor  1.000141e+03  1.478197e+06   
lr                      Linear Regression  1.651466e+15  7.275465e+30   
lar                Least Angle Regression  2.026117e+21  8.805689e+43   

                  RMSE            R2    RMSLE          MAPE  TT (Sec)  
omp       3.080031e+02  8.155000e-01   0.0894  6.460000e-02     0.086  
gbr       3.303561e+02  7.167000e-01   0.1042  8.790000e-02     0.568  
xgboost   3.499655e+02  6.944000e-01   0.1108  9.390000e-02     0.387  
lasso     3.884556e+02  6.748000e-01   0.1212  1.042000e-01     0.095  
rf        3.694456e+02  6.698000e-01   0.1140  9.600000e-02     0.583  
llar      3.936485e+02  6.664000e-01   0.1234  1.066000e-01     0.090  
et        4.064558e+02  6.556000e-01   0.1316  1.129000e-01     0.402  
ridge     4.041292e+02  6.460000e-01   0.1239  1.072000e-01     0.107  
br        4.044734e+02  6.453000e-01   0.1240  1.074000e-01     0.093  
lightgbm  3.751441e+02  6.397000e-01   0.1134  9.490000e-02     0.902  
ada       4.169265e+02  6.144000e-01   0.1307  1.132000e-01     0.311  
dt        4.333035e+02  5.785000e-01   0.1345  1.121000e-01     0.104  
par       4.152513e+02  5.651000e-01   0.1270  1.093000e-01     0.088  
huber     4.242954e+02  5.126000e-01   0.1293  1.116000e-01     0.125  
en        6.239081e+02  2.786000e-01   0.1924  1.744000e-01     0.089  
knn       7.141765e+02  5.800000e-03   0.2146  1.926000e-01     0.098  
dummy     1.173618e+03 -1.344800e+00   0.3538  3.398000e-01     0.081  
lr        2.070237e+15 -1.506688e+25  17.9285  5.201707e+11     0.092  
lar       2.967440e+21 -1.487704e+38  10.2284  4.499684e+17     0.096

In [7]:
# Extraer resultados
results = pull()

# Mostrar tabla en decimales normales, redondeada a 2
pd.options.display.float_format = '{:.4f}'.format

# Ver resultados
results

Model                         MAE  \
omp           Orthogonal Matching Pursuit                    221.4173   
gbr           Gradient Boosting Regressor                    259.2703   
xgboost         Extreme Gradient Boosting                    275.0347   
lasso                    Lasso Regression                    310.2573   
rf                Random Forest Regressor                    285.4094   
llar         Lasso Least Angle Regression                    314.8613   
et                  Extra Trees Regressor                    329.8491   
ridge                    Ridge Regression                    327.3401   
br                         Bayesian Ridge                    327.6265   
lightgbm  Light Gradient Boosting Machine                    284.9510   
ada                    AdaBoost Regressor                    332.3053   
dt                Decision Tree Regressor                    343.6374   
par          Passive Aggressive Regressor                    332.4782   
huber                     Huber Regressor                    338.7426   
en                            Elastic Net                    534.7201   
knn                 K Neighbors Regressor                    614.4785   
dummy                     Dummy Regressor                   1000.1410   
lr                      Linear Regression       1651465705290255.7500   
lar                Least Angle Regression 2026117020779660443648.0000   

                                                       MSE  \
omp                                            106194.0265   
gbr                                            146062.7920   
xgboost                                        156559.7213   
lasso                                          167530.5090   
rf                                             172591.4777   
llar                                           172581.7840   
et                                             195728.1268   
ridge                                          182064.9330   
br                                             182475.5309   
lightgbm                                       195911.0251   
ada                                            202736.4867   
dt                                             225087.4096   
par                                            192604.6385   
huber                                          201097.4394   
en                                             441964.2988   
knn                                            602631.0523   
dummy                                         1478197.0635   
lr                    7275464567888223682491641233408.0000   
lar      88056890581607712035261580796348349431152640.0000   

                                RMSE  \
omp                         308.0031   
gbr                         330.3561   
xgboost                     349.9655   
lasso                       388.4556   
rf                          369.4456   
llar                        393.6485   
et                          406.4558   
ridge                       404.1292   
br                          404.4734   
lightgbm                    375.1441   
ada                         416.9265   
dt                          433.3035   
par                         415.2513   
huber                       424.2954   
en                          623.9081   
knn                         714.1765   
dummy                      1173.6183   
lr             2070237075545042.0000   
lar      2967440145160433303552.0000   

                                                    R2   RMSLE  \
omp                                             0.8155  0.0894   
gbr                                             0.7167  0.1042   
xgboost                                         0.6944  0.1108   
lasso                                           0.6748  0.1212   
rf                                              0.6698  0.1140   
llar                                            0.6664  0.1234   
et                                              0.6556  0.1316   
ridge

### Ver si mejoran las métricas con XGBoost

In [11]:
from pycaret.regression import setup, compare_models, pull, stack_models, blend_models, evaluate_model

# Setup inicial
exp = setup(
    data=df_pycaret,
    target="PRECIO M2",
    session_id=123,
    normalize=True,
    remove_multicollinearity=True,
    multicollinearity_threshold=0.8,
    imputation_type="simple",
    numeric_imputation="median",
    categorical_imputation="mode",
    use_gpu=True,
    fold_strategy="timeseries",
    data_split_shuffle=False,   
    fold_shuffle=False
)

# Comparar modelos (ahora debería aparecer también XGBoost si se instaló bien)
top_models = compare_models(n_select=5, sort="R2")

[LightGBM] [Warning] There are no meaningful features which satisfy the provided configuration. Decreasing Dataset parameters min_data_in_bin or min_data_in_leaf and re-constructing Dataset might resolve this warning.
[LightGBM] [Info] This is the GPU trainer!!
[LightGBM] [Info] Total Bins 0
[LightGBM] [Info] Number of data points in the train set: 2, number of used features: 0
[LightGBM] [Info] Using GPU Device: NVIDIA GeForce RTX 3050 OEM, Vendor: NVIDIA Corporation
[LightGBM] [Info] Compiling OpenCL Kernel with 16 bins...
[LightGBM] [Info] GPU programs have been built
[LightGBM] [Warning] GPU acceleration is disabled because no non-trivial dense features can be found
[LightGBM] [Info] Start training from score 0.500000
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[LightGBM] [Warning] Stopped training because there are no m

[LightGBM] [Warning] There are no meaningful features which satisfy the provided configuration. Decreasing Dataset parameters min_data_in_bin or min_data_in_leaf and re-constructing Dataset might resolve this warning.
[LightGBM] [Info] This is the GPU trainer!!
[LightGBM] [Info] Total Bins 0
[LightGBM] [Info] Number of data points in the train set: 2, number of used features: 0
[LightGBM] [Info] Using GPU Device: NVIDIA GeForce RTX 3050 OEM, Vendor: NVIDIA Corporation
[LightGBM] [Info] Compiling OpenCL Kernel with 16 bins...
[LightGBM] [Info] GPU programs have been built
[LightGBM] [Warning] GPU acceleration is disabled because no non-trivial dense features can be found
[LightGBM] [Info] Start training from score 0.500000
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[LightGBM] [Warning] Stopped training because there are no m

In [12]:
# Stacking: combina varios modelos en un meta-modelo
stacked_model = stack_models(
    estimator_list=top_models,   # lista de mejores modelos
    meta_model=None,             # si lo dejas en None, usa LinearRegression por defecto
    fold=5,                      # validación cruzada
    optimize="R2"
)

# Evaluar el modelo apilado
evaluate_model(stacked_model)


interactive(children=(ToggleButtons(description='Plot Type:', icons=('',), options=(('Pipeline Plot', 'pipelin…

In [13]:
blended_model = blend_models(
    estimator_list=top_models,
    fold=5,
    optimize="R2"
)

evaluate_model(blended_model)


interactive(children=(ToggleButtons(description='Plot Type:', icons=('',), options=(('Pipeline Plot', 'pipelin…

### 2. PYCARET imputando nulos y sin One Hot Encoding

In [14]:
df_pycaret1 = pd.read_excel(r"C:\Users\evahr\Downloads\TFM-idealista\dataset_entrenamiento_lag.xlsx")

df_pycaret1

AÑO  MES_NUM    DISTRITO  LATITUD  LONGITUD    SHAPE_AREA  PRECIO M2  \
0     2008       11  Arganzuela  40.4002   -3.6962  6462175.9140       3714   
1     2008       12  Arganzuela  40.4002   -3.6962  6462175.9140       3667   
2     2009        1  Arganzuela  40.4002   -3.6962  6462175.9140       3782   
3     2009        2  Arganzuela  40.4002   -3.6962  6462175.9140       3722   
4     2009        3  Arganzuela  40.4002   -3.6962  6462175.9140       3722   
...    ...      ...         ...      ...       ...           ...        ...   
4161  2024       12  Villaverde  40.3500   -3.7000 20194322.8326       2252   
4162  2025        1  Villaverde  40.3500   -3.7000 20194322.8326       2207   
4163  2025        2  Villaverde  40.3500   -3.7000 20194322.8326       2254   
4164  2025        3  Villaverde  40.3500   -3.7000 20194322.8326       2294   
4165  2025        4  Villaverde  40.3500   -3.7000 20194322.8326       2362   

      EURIBOR_lag12  VARIACIÓN ANUAL_lag12  PRECIO M2_lag12  \
0            0.0461                 0.0125             4000   
1            0.0479                 0.0125             4000   
2            0.0450                 0.0125             4000   
3            0.0435                 0.0125             3824   
4            0.0459                 0.0125             4000   
...             ...                    ...              ...   
4161         0.0368                 0.0630             1963   
4162         0.0361                 0.0680             1982   
4163         0.0367                 0.0580             1977   
4164         0.0372                 0.0670             1988   
4165         0.0370                 0.0850             2046   

      TRANSACCIONES_CM_lag12  INDICE_PRECIO_lag12  TOTAL_HIPOTECAS_lag12  \
0                       6058             154.5100                  13651   
1                       4577             154.5100                   9041   
2                       6123             152.9000                  13884   
3                       5964             152.9000                  13536   
4                       5525             152.9000                  12024   
...                      ...                  ...                    ...   
4161                    4048             164.8200                   6548   
4162                    6209             168.7900                   8547   
4163                    6336             168.7900                   8492   
4164                    5924             168.7900                   7033   
4165                    5805             174.5400                   7177   

      POBLACION_ACTIVA_lag12  POBLACION_lag12  ESPERANZA_VIDA_lag12  \
0                  3365.3000           153868               83.4000   
1                  3365.3000           153218               83.4000   
2                  3375.6000           153670               83.3000   
3                  3375.6000           153794               83.3000   
4                  3375.6000           154370               83.3000   
...                      ...              ...                   ...   
4161               3720.2000           318076               85.9000   
4162               3750.8000           335904               88.8000   
4163               3750.8000           335904               88.8000   
4164               3750.8000           335904               88.8000   
4165               3749.9000           335904               88.8000   

      VIVIENDAS_COMPRAVENTA_lag12  TRANSACCIONES_SUELO_lag12  \
0                            2539                        464   
1                            1819                        464   
2                            2278                        458   
3                            2347                        458   
4                            2507                        458   
...                           ...                        ...   
4161                          815                        345   
4162                   

In [15]:
setup(data=df_pycaret1, target='PRECIO M2', preprocess=False, session_id=123)

In [16]:
from pycaret.regression import setup, compare_models, pull

# Setup inicial
exp = setup(
    data=df_pycaret1,
    target="PRECIO M2",
    session_id=123,
    normalize=True,
    remove_multicollinearity=True,
    multicollinearity_threshold=0.8,
    imputation_type="simple",
    numeric_imputation="median",
    categorical_imputation="mode",
    use_gpu=True,
    fold_strategy="timeseries",
    data_split_shuffle=False,   
    fold_shuffle=False
)

# Comparación de modelos
best_models1 = compare_models(n_select=10, sort="R2")

[LightGBM] [Warning] There are no meaningful features which satisfy the provided configuration. Decreasing Dataset parameters min_data_in_bin or min_data_in_leaf and re-constructing Dataset might resolve this warning.
[LightGBM] [Info] This is the GPU trainer!!
[LightGBM] [Info] Total Bins 0
[LightGBM] [Info] Number of data points in the train set: 2, number of used features: 0
[LightGBM] [Info] Using GPU Device: NVIDIA GeForce RTX 3050 OEM, Vendor: NVIDIA Corporation
[LightGBM] [Info] Compiling OpenCL Kernel with 16 bins...
[LightGBM] [Info] GPU programs have been built
[LightGBM] [Warning] GPU acceleration is disabled because no non-trivial dense features can be found
[LightGBM] [Info] Start training from score 0.500000
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[LightGBM] [Warning] Stopped training because there are no m

[LightGBM] [Warning] There are no meaningful features which satisfy the provided configuration. Decreasing Dataset parameters min_data_in_bin or min_data_in_leaf and re-constructing Dataset might resolve this warning.
[LightGBM] [Info] This is the GPU trainer!!
[LightGBM] [Info] Total Bins 0
[LightGBM] [Info] Number of data points in the train set: 2, number of used features: 0
[LightGBM] [Info] Using GPU Device: NVIDIA GeForce RTX 3050 OEM, Vendor: NVIDIA Corporation
[LightGBM] [Info] Compiling OpenCL Kernel with 16 bins...
[LightGBM] [Info] GPU programs have been built
[LightGBM] [Warning] GPU acceleration is disabled because no non-trivial dense features can be found
[LightGBM] [Info] Start training from score 0.500000
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[LightGBM] [Warning] Stopped training because there are no m

In [17]:
# Extraer resultados de la última operación (compare_models)
results = pull()

# Mostrar tabla (DataFrame con R2, RMSE, MAE, etc.)
results

Model                   MAE  \
omp           Orthogonal Matching Pursuit              178.2331   
gbr           Gradient Boosting Regressor              207.5812   
lasso                    Lasso Regression              223.6960   
llar         Lasso Least Angle Regression              223.9263   
xgboost         Extreme Gradient Boosting              245.1793   
ridge                    Ridge Regression              244.1111   
br                         Bayesian Ridge              243.9519   
rf                Random Forest Regressor              254.2759   
et                  Extra Trees Regressor              301.5554   
ada                    AdaBoost Regressor              268.3121   
lightgbm  Light Gradient Boosting Machine              260.5657   
dt                Decision Tree Regressor              307.2358   
par          Passive Aggressive Regressor              281.2053   
huber                     Huber Regressor              278.8877   
en                            Elastic Net              539.9834   
knn                 K Neighbors Regressor              615.2642   
dummy                     Dummy Regressor             1014.6102   
lr                      Linear Regression 2122210648383256.2500   
lar                Least Angle Regression 8902375328576214.0000   

                                            MSE                   RMSE  \
omp                                  56486.1840               227.8842   
gbr                                 109429.0601               260.8409   
lasso                               101866.9116               283.2888   
llar                                102042.4844               283.5422   
xgboost                             133440.5541               298.6806   
ridge                               117369.2019               304.5486   
br                                  117457.0471               304.4402   
rf                                  142556.2072               312.6237   
et                                  171340.9392               361.8766   
ada                                 155074.6022               331.4399   
lightgbm                            161980.9277               322.5162   
dt                                  182535.9642               376.7640   
par                                 144827.8498               345.9836   
huber                               146018.7362               344.8463   
en                                  442462.0229               626.4601   
knn                                 623744.3823               717.1659   
dummy                              1524396.0934              1192.0622   
lr        11416852324701946640733541761024.0000  2590577955046782.5000   
lar      508125838414196433343940158029824.0000 10844924906895858.0000   

                                         R2   RMSLE               MAPE  \
omp                                  0.8996  0.0670             0.0528   
gbr                                  0.7815  0.0841             0.0735   
lasso                                0.7636  0.0878             0.0758   
llar                                 0.7630  0.0878             0.0759   
xgboost                              0.7264  0.0943             0.0855   
ridge                                0.7190  0.0937             0.0817   
br                                   0.7166  0.0938             0.0818   
rf                                   0.7094  0.0990             0.0886   
et                                   0.6952  0.1189             0.1060   
ada                                  0.6893  0.1091             0.0976   
lightgbm                             0.6837  0.0992             0.0893   
dt                                   0.6223  0.1150             0.1029   
par                                  0.6079  0.1051             0.0939   
huber                                0.6000  0.1056             0.0937   
en                                   0.3034  0.1929             0.1767   
knn                           

In [18]:

# Extraer resultados
results = pull()

# Mostrar tabla en decimales normales, redondeada a 2
pd.options.display.float_format = '{:.4f}'.format

# Ver resultados
results


Model                   MAE  \
omp           Orthogonal Matching Pursuit              178.2331   
gbr           Gradient Boosting Regressor              207.5812   
lasso                    Lasso Regression              223.6960   
llar         Lasso Least Angle Regression              223.9263   
xgboost         Extreme Gradient Boosting              245.1793   
ridge                    Ridge Regression              244.1111   
br                         Bayesian Ridge              243.9519   
rf                Random Forest Regressor              254.2759   
et                  Extra Trees Regressor              301.5554   
ada                    AdaBoost Regressor              268.3121   
lightgbm  Light Gradient Boosting Machine              260.5657   
dt                Decision Tree Regressor              307.2358   
par          Passive Aggressive Regressor              281.2053   
huber                     Huber Regressor              278.8877   
en                            Elastic Net              539.9834   
knn                 K Neighbors Regressor              615.2642   
dummy                     Dummy Regressor             1014.6102   
lr                      Linear Regression 2122210648383256.2500   
lar                Least Angle Regression 8902375328576214.0000   

                                            MSE                   RMSE  \
omp                                  56486.1840               227.8842   
gbr                                 109429.0601               260.8409   
lasso                               101866.9116               283.2888   
llar                                102042.4844               283.5422   
xgboost                             133440.5541               298.6806   
ridge                               117369.2019               304.5486   
br                                  117457.0471               304.4402   
rf                                  142556.2072               312.6237   
et                                  171340.9392               361.8766   
ada                                 155074.6022               331.4399   
lightgbm                            161980.9277               322.5162   
dt                                  182535.9642               376.7640   
par                                 144827.8498               345.9836   
huber                               146018.7362               344.8463   
en                                  442462.0229               626.4601   
knn                                 623744.3823               717.1659   
dummy                              1524396.0934              1192.0622   
lr        11416852324701946640733541761024.0000  2590577955046782.5000   
lar      508125838414196433343940158029824.0000 10844924906895858.0000   

                                         R2   RMSLE               MAPE  \
omp                                  0.8996  0.0670             0.0528   
gbr                                  0.7815  0.0841             0.0735   
lasso                                0.7636  0.0878             0.0758   
llar                                 0.7630  0.0878             0.0759   
xgboost                              0.7264  0.0943             0.0855   
ridge                                0.7190  0.0937             0.0817   
br                                   0.7166  0.0938             0.0818   
rf                                   0.7094  0.0990             0.0886   
et                                   0.6952  0.1189             0.1060   
ada                                  0.6893  0.1091             0.0976   
lightgbm                             0.6837  0.0992             0.0893   
dt                                   0.6223  0.1150             0.1029   
par                                  0.6079  0.1051             0.0939   
huber                                0.6000  0.1056             0.0937   
en                                   0.3034  0.1929             0.1767   
knn                           

### 3. PYCARET imputando nulos y con One Hot Encoding -> resultados seleccionados

In [19]:
df_pycaret2 = pd.read_excel(r"C:\Users\evahr\Downloads\TFM-idealista\dataset_entrenamiento_lag1.xlsx")

df_pycaret2

AÑO  MES_NUM  LATITUD  LONGITUD    SHAPE_AREA  PRECIO M2  \
0     2008       11  40.4002   -3.6962  6462175.9140       3714   
1     2008       12  40.4002   -3.6962  6462175.9140       3667   
2     2009        1  40.4002   -3.6962  6462175.9140       3782   
3     2009        2  40.4002   -3.6962  6462175.9140       3722   
4     2009        3  40.4002   -3.6962  6462175.9140       3722   
...    ...      ...      ...       ...           ...        ...   
4161  2024       12  40.3500   -3.7000 20194322.8326       2252   
4162  2025        1  40.3500   -3.7000 20194322.8326       2207   
4163  2025        2  40.3500   -3.7000 20194322.8326       2254   
4164  2025        3  40.3500   -3.7000 20194322.8326       2294   
4165  2025        4  40.3500   -3.7000 20194322.8326       2362   

      EURIBOR_lag12  VARIACIÓN ANUAL_lag12  PRECIO M2_lag12  \
0            0.0461                 0.0125             4000   
1            0.0479                 0.0125             4000   
2            0.0450                 0.0125             4000   
3            0.0435                 0.0125             3824   
4            0.0459                 0.0125             4000   
...             ...                    ...              ...   
4161         0.0368                 0.0630             1963   
4162         0.0361                 0.0680             1982   
4163         0.0367                 0.0580             1977   
4164         0.0372                 0.0670             1988   
4165         0.0370                 0.0850             2046   

      TRANSACCIONES_CM_lag12  ...  DIST_Moratalaz  DIST_Puente De Vallecas  \
0                       6058  ...           False                    False   
1                       4577  ...           False                    False   
2                       6123  ...           False                    False   
3                       5964  ...           False                    False   
4                       5525  ...           False                    False   
...                      ...  ...             ...                      ...   
4161                    4048  ...           False                    False   
4162                    6209  ...           False                    False   
4163                    6336  ...           False                    False   
4164                    5924  ...           False                    False   
4165                    5805  ...           False                    False   

      DIST_Retiro  DIST_Salamanca  DIST_San Blas-Canillejas  DIST_Tetuan  \
0           False           False                     False        False   
1           False           False                     False        False   
2           False           False                     False        False   
3           False           False                     False        False   
4           False           False                     False        False   
...           ...             ...                       ...          ...   
4161        False           False                     False        False   
4162        False           False                     False        False   
4163        False           False                     False        False   
4164        False           False                     False        False   
4165        False           False                     False        False   

      DIST_Usera  DIST_Vicalvaro  DIST_Villa De Vallecas  DIST_Villaverde  
0          False           False                   False            False  
1          False           False                   False            False  
2          False           False                   False            False  
3          False           False                   False            False  
4          False           False                   False            False  
...          ...             ...                     ...              ...  
4161       False           False                   False       

In [20]:
setup(data=df_pycaret2, target='PRECIO M2', preprocess=False, session_id=123)

In [21]:
from pycaret.regression import setup, compare_models, pull

# Setup inicial
exp = setup(
    data=df_pycaret2,
    target="PRECIO M2",
    session_id=123,
    normalize=True,
    remove_multicollinearity=True,
    multicollinearity_threshold=0.8,
    imputation_type="simple",
    numeric_imputation="median",
    categorical_imputation="mode",
    use_gpu=True,
    fold_strategy="timeseries",
    data_split_shuffle=False,   
    fold_shuffle=False
)

# Comparación de modelos
best_models2 = compare_models(n_select=10, sort="R2")

[LightGBM] [Warning] There are no meaningful features which satisfy the provided configuration. Decreasing Dataset parameters min_data_in_bin or min_data_in_leaf and re-constructing Dataset might resolve this warning.
[LightGBM] [Info] This is the GPU trainer!!
[LightGBM] [Info] Total Bins 0
[LightGBM] [Info] Number of data points in the train set: 2, number of used features: 0
[LightGBM] [Info] Using GPU Device: NVIDIA GeForce RTX 3050 OEM, Vendor: NVIDIA Corporation
[LightGBM] [Info] Compiling OpenCL Kernel with 16 bins...
[LightGBM] [Info] GPU programs have been built
[LightGBM] [Warning] GPU acceleration is disabled because no non-trivial dense features can be found
[LightGBM] [Info] Start training from score 0.500000
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[LightGBM] [Warning] Stopped training because there are no m

[LightGBM] [Warning] There are no meaningful features which satisfy the provided configuration. Decreasing Dataset parameters min_data_in_bin or min_data_in_leaf and re-constructing Dataset might resolve this warning.
[LightGBM] [Info] This is the GPU trainer!!
[LightGBM] [Info] Total Bins 0
[LightGBM] [Info] Number of data points in the train set: 2, number of used features: 0
[LightGBM] [Info] Using GPU Device: NVIDIA GeForce RTX 3050 OEM, Vendor: NVIDIA Corporation
[LightGBM] [Info] Compiling OpenCL Kernel with 16 bins...
[LightGBM] [Info] GPU programs have been built
[LightGBM] [Warning] GPU acceleration is disabled because no non-trivial dense features can be found
[LightGBM] [Info] Start training from score 0.500000
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[LightGBM] [Warning] Stopped training because there are no m

In [22]:
# Extraer resultados de la última operación (compare_models)
results = pull()

# Mostrar tabla (DataFrame con R2, RMSE, MAE, etc.)
results

Model                    MAE  \
omp           Orthogonal Matching Pursuit               155.5967   
gbr           Gradient Boosting Regressor               204.1255   
llar         Lasso Least Angle Regression               223.9263   
lasso                    Lasso Regression               224.2709   
xgboost         Extreme Gradient Boosting               244.4361   
ridge                    Ridge Regression               244.1111   
br                         Bayesian Ridge               243.9519   
rf                Random Forest Regressor               253.9741   
ada                    AdaBoost Regressor               267.7679   
lightgbm  Light Gradient Boosting Machine               259.5999   
et                  Extra Trees Regressor               307.0811   
dt                Decision Tree Regressor               312.9283   
par          Passive Aggressive Regressor               281.2053   
huber                     Huber Regressor               278.8877   
en                            Elastic Net               539.9772   
knn                 K Neighbors Regressor               615.2642   
dummy                     Dummy Regressor              1014.6102   
lr                      Linear Regression  1185512888317047.5000   
lar                Least Angle Regression 47225823194891760.0000   

                                              MSE                   RMSE  \
omp                                    43321.7545               201.3818   
gbr                                   106484.1988               257.2093   
llar                                  102042.4844               283.5422   
lasso                                 102108.9394               283.8017   
xgboost                               133390.2505               298.5664   
ridge                                 117369.2019               304.5486   
br                                    117457.0471               304.4402   
rf                                    141768.2940               311.7246   
ada                                   153565.1463               330.8401   
lightgbm                              161459.6884               321.9082   
et                                    187091.1323               369.0037   
dt                                    181973.3781               379.6609   
par                                   144827.8498               345.9836   
huber                                 146018.7362               344.8463   
en                                    442453.9822               626.4550   
knn                                   623744.3823               717.1659   
dummy                                1524396.0934              1192.0622   
lr           5495430546698183635658250125312.0000  1577907821527598.5000   
lar      36498784142201066779527238752665600.0000 70666534294961128.0000   

                                          R2   RMSLE                MAPE  \
omp                                   0.9183  0.0576              0.0466   
gbr                                   0.7847  0.0834              0.0727   
llar                                  0.7630  0.0878              0.0759   
lasso                                 0.7627  0.0880              0.0761   
xgboost                               0.7263  0.0941              0.0852   
ridge                                 0.7190  0.0937              0.0817   
br                                    0.7166  0.0938              0.0818   
rf                                    0.7117  0.0991              0.0887   
ada                                   0.6913  0.1092              0.0975   
lightgbm                              0.6842  0.0989              0.0889   
et                                    0.6596  0.1207              0.1079   
dt                                    0.6387  0.1151              0.1038   
par                                   0.6079  0.1051              0.0939   
huber                                 0.6000  0.1056              0.0937   
en            

In [23]:
# Extraer resultados
results = pull()

# Mostrar tabla en decimales normales, redondeada a 2
pd.options.display.float_format = '{:.4f}'.format

# Ver resultados
results

Model                    MAE  \
omp           Orthogonal Matching Pursuit               155.5967   
gbr           Gradient Boosting Regressor               204.1255   
llar         Lasso Least Angle Regression               223.9263   
lasso                    Lasso Regression               224.2709   
xgboost         Extreme Gradient Boosting               244.4361   
ridge                    Ridge Regression               244.1111   
br                         Bayesian Ridge               243.9519   
rf                Random Forest Regressor               253.9741   
ada                    AdaBoost Regressor               267.7679   
lightgbm  Light Gradient Boosting Machine               259.5999   
et                  Extra Trees Regressor               307.0811   
dt                Decision Tree Regressor               312.9283   
par          Passive Aggressive Regressor               281.2053   
huber                     Huber Regressor               278.8877   
en                            Elastic Net               539.9772   
knn                 K Neighbors Regressor               615.2642   
dummy                     Dummy Regressor              1014.6102   
lr                      Linear Regression  1185512888317047.5000   
lar                Least Angle Regression 47225823194891760.0000   

                                              MSE                   RMSE  \
omp                                    43321.7545               201.3818   
gbr                                   106484.1988               257.2093   
llar                                  102042.4844               283.5422   
lasso                                 102108.9394               283.8017   
xgboost                               133390.2505               298.5664   
ridge                                 117369.2019               304.5486   
br                                    117457.0471               304.4402   
rf                                    141768.2940               311.7246   
ada                                   153565.1463               330.8401   
lightgbm                              161459.6884               321.9082   
et                                    187091.1323               369.0037   
dt                                    181973.3781               379.6609   
par                                   144827.8498               345.9836   
huber                                 146018.7362               344.8463   
en                                    442453.9822               626.4550   
knn                                   623744.3823               717.1659   
dummy                                1524396.0934              1192.0622   
lr           5495430546698183635658250125312.0000  1577907821527598.5000   
lar      36498784142201066779527238752665600.0000 70666534294961128.0000   

                                          R2   RMSLE                MAPE  \
omp                                   0.9183  0.0576              0.0466   
gbr                                   0.7847  0.0834              0.0727   
llar                                  0.7630  0.0878              0.0759   
lasso                                 0.7627  0.0880              0.0761   
xgboost                               0.7263  0.0941              0.0852   
ridge                                 0.7190  0.0937              0.0817   
br                                    0.7166  0.0938              0.0818   
rf                                    0.7117  0.0991              0.0887   
ada                                   0.6913  0.1092              0.0975   
lightgbm                              0.6842  0.0989              0.0889   
et                                    0.6596  0.1207              0.1079   
dt                                    0.6387  0.1151              0.1038   
par                                   0.6079  0.1051              0.0939   
huber                                 0.6000  0.1056              0.0937   
en            

In [28]:
results.to_excel(r"C:\Users\evahr\Downloads\TFM-idealista\resultados_modelos.xlsx", index=False, float_format="%.4f")

Se aprecia una mejora significativa al aplicar OHE.

La comparativa de algoritmos mostró que los modelos lineales ofrecen el mejor rendimiento en este conjunto de datos.
El modelo Orthogonal Matching Pursuit (OMP) alcanzó un R² de 0.9183 con un RMSE de 201 €/m², siendo el más preciso.

Otros modelos lineales como Lasso, LassoLars, Ridge y Bayesian Ridge obtuvieron un desempeño ligeramente inferior, con valores de R² entre 0.71 y 0.76 y RMSE entre 280 y 305 €/m².

En contraste, los métodos de ensamblado basados en árboles (Random Forest, Gradient Boosting, XGBoost, LightGBM y Extra Trees) mostraron un rendimiento más bajo, con R² ≤ 0.78 y RMSE por encima de 250 €/m².

Finalmente, modelos más complejos como Elastic Net, KNN, Passive Aggressive o Huber Regressor tampoco lograron mejorar el desempeño, quedándose con R² entre 0.30 y 0.60 y errores significativamente mayores.

Estos resultados confirman que, en este dataset, la relación entre las variables explicativas y el precio por m² se ajusta mejor a un modelo de tipo lineal con codificación adecuada de variables categóricas, que a estructuras no lineales más complejas como los ensamblados.

🔝 Mejores modelos según R² y error

Orthogonal Matching Pursuit (omp) →

R² = 0.9183, el más alto de todos los modelos.

RMSE ≈ 201 €/m², MAE ≈ 156 €/m² → errores significativamente más bajos que el resto.
✅ Es el mejor modelo: combina precisión y simplicidad, mostrando que la relación entre las variables y el precio/m² es casi lineal.

Gradient Boosting Regressor (gbr) →

R² = 0.7847, bastante inferior a OMP.

RMSE ≈ 257 €/m², MAE ≈ 204 €/m².
Un modelo aceptable, pero con más error y menor capacidad explicativa.

Lasso, LassoLars, Ridge, Bayesian Ridge →

R² entre 0.71 y 0.76.

RMSE entre 280 y 305 €/m².
Modelos lineales regulares, más interpretables, pero claramente menos precisos que OMP.

📊 Modelos de ensamblado (XGBoost, Random Forest, LightGBM, Extra Trees)

R² entre 0.65 y 0.72, lo que indica que no capturan bien la estructura de los datos en este escenario.

RMSE entre 298 y 369 €/m², bastante superiores a los modelos lineales ajustados con OHE.
⚠️ Contrario a lo esperado, aquí los árboles no aportan ventaja.

⚖️ Otros modelos

Elastic Net, Passive Aggressive, Huber, KNN, AdaBoost, Decision Tree: rendimiento intermedio o bajo, con R² ≤ 0.60 y errores altos.

Linear Regression y Least Angle Regression → se desestabilizan y dan resultados inválidos (valores negativos de R²).

Dummy Regressor → R² ≈ 0, básicamente predice la media.

⏱️ Tiempos de entrenamiento (TT Sec)

Todos los modelos entrenan en menos de 1 segundo.

OMP es extremadamente rápido y a la vez el más preciso.

✅ Conclusión práctica
Los datos, tras imputación y one hot encoding, muestran un patrón altamente lineal.
El mejor modelo es Orthogonal Matching Pursuit, que logra el equilibrio óptimo entre precisión, interpretabilidad y eficiencia.
Los modelos de ensamblado, que en otros contextos suelen dominar, aquí no superan a los lineales y muestran un ajuste bastante peor.